File: server/database/models/user/user_model.ts
import jwt from "jsonwebtoken";
import { Document, model, Schema } from "mongoose";

//!
//! INITIAL SCHEMA / PAYLOAD OR DATA STRUCTURE FOR OUR AUTHORIZED USER
//! AS IT SHOULD BE STORED ON THE BE
const UserSchema = new Schema({
  name: {
    type: String,
    required: [true, "Please enter your name"],
    trim: true,
    maxlength: [30, "Your name cannot exceed 30 characters"],
    minlength: [3, "Your name must be at least 3 characters long"],
  },
  email: {
    type: String,
    required: [true, "Please enter your email"],
    trim: true,
    unique: true,
    match: [
      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
      "Please provide a valid email",
    ],
  },
  profilePic: {
    type: String,
    required: true,
  },
});

//!
//! CREATING JWT
UserSchema.methods.createJWT = function (uuid: string): string {
  const token = jwt.sign(
    { userId: this._id, email: this.email },
    process.env.JWT_SECRET,
    {
      expiresIn: process.env.JWT_EXPIRES,
      issuer: process.env.JWT_ISSUER,
      jwtid: uuid,
    }
  );
  return token;
};

//!
//! CREATING REFRESH TOKENS
UserSchema.methods.createRefresh = function (uuid: string): string {
  const refreshToken = jwt.sign(
    { userId: this._id, email: this.email },
    process.env.REFRESH_SECRET,
    {
      expiresIn: process.env.REFRESH_EXPIRES,
      issuer: process.env.JWT_ISSUER,
      jwtid: uuid,
    }
  );
  return refreshToken;
};

//!
//! USER SCHEMA TO JSON
UserSchema.methods.toJSON = function (): any {
  return {
    id: this._id,
    name: this.name,
    email: this.email,
    profilePic: this.profilePic,
  };
};

//!
//! CREATING MONGOOSE - MONGODB DOCUMENT
//! ACTUAL SCHEMA
export interface UserDocument extends Document {
  name: string;
  email: string;
  profilePic: string;
  createJWT: (uuid: string) => string;
  createRefresh: (uuid: string) => string;

  toJSON: () => any;
}

export default model<UserDocument>("User", UserSchema);

File: lib/services/models/user_model.dart
import "dart:collection";

class UserModel extends MapView<String, dynamic> {
  UserModel({
    required this.email,
    required this.name,
    required this.profilePic,
    required this.userID,
    required this.token,
  }) : super({
          "userID": userID,
          "email": email,
          "name": name,
          "profilePic": profilePic,
          "token": token,
        });

  factory UserModel.fromJSON({
    required Map<String, dynamic> json,
  }) {
    final Map<String, dynamic> userDOC = json["user"];

    return UserModel(
      email: userDOC["email"] ?? "",
      name: userDOC["name"] ?? "",
      profilePic: userDOC["profilePic"] ?? "",
      userID: userDOC["id"] ?? "",
      token: json["token"] ?? "",
    );
  }

  final String email;
  final String name;
  final String profilePic;
  final String userID;
  final String token;

  UserModel copyWith({
    String? email,
    String? name,
    String? profilePic,
    String? userID,
    String? token,
  }) {
    return UserModel(
      email: email ?? this.email,
      name: name ?? this.name,
      profilePic: profilePic ?? this.profilePic,
      userID: userID ?? this.userID,
      token: token ?? this.token,
    );
  }
}
